File: src/main/Python/boj_step16/boj4949.py
def isVPS(s):
    VPS = []
    for c in s:
        if c == ')' and (len(VPS) == 0 or not VPS.pop()): return False
        if c == ']' and (len(VPS) == 0 or VPS.pop()): return False
        if c == '(': VPS.append(True)
        if c == '[': VPS.append(False)
    if len(VPS) == 0: return True
    else: return False

while True:
    s = input()
    if s == '.': exit()
    print('yes' if isVPS(s) else 'no')

File: src/main/Cpp/boj_step16/boj4949.cpp
#include <iostream>
#include <stack>
#include <functional>
using namespace std;

int main() {
    function isVPS = [&](stack<char> st, string s) -> string {
        for (const auto & c : s) {
            if (c == '(' || c == '[')
                st.push(c);
            else if (c == ')') {
                if (st.empty() || st.top() == '[')
                    return "no";

                st.pop();
            } else if (c == ']') {
                if (st.empty() || st.top()== '(')
                    return "no";

                st.pop();
            }
        }

        return (st.empty()) ? "yes" : "no";
    };

    string s;
    stack<char> st;
    while (getline(cin, s) && s != ".") {
        cout << isVPS(st, s) << endl;
    }
}